src/Deinterleave.cpp

```diff
@@ -176,224 +178,647 @@ Stmt collect_strided_stores(const Stmt &stmt, const std::string &name, int strid
+    int min_difference = std::numeric_limits<int>::max();
+    std::adjacent_find(indices.begin(), indices.end(),
+                       [&min_difference](int a, int b) {
+                           min_difference = std::min(min_difference, b - a);
+                           return false;
+                       });
+    return min_difference;
+}
+
+// Which 1-st order polynominal goes through all of the specified \p indices,
+// while going through the minimal amount of the indices *NOT* specified here?
+std::optional<std::pair<int /*base*/, int /*min_step*/>>
+fit_affine(const std::vector<int> &indices) {
+    if (indices.empty()) {
+        return {};
+    }
+    int base = indices.front();
+    int min_step =
+        indices.size() < 2 ? 1 : minimal_adjacent_difference(indices);
+    return std::make_pair(base, min_step);
+}
+
+// Return the 1-st order polynominal that exactly specifies \p indices.
+// Note: there might not be such a polynominal.
+std::optional<std::pair<int /*base*/, int /*step*/>>
+as_affine(const std::vector<int> &indices) {
+    if (indices.empty()) {
+        return {};
+    }
+    int base = indices.front();
+    int step = indices.size() < 2 ? 1 : indices[1] - indices[0];
+    auto step_mismatch = [step](int a, int b) {
+        int curr_step = b - a;
+        return curr_step != step;
+    };
+    if (std::adjacent_find(std::next(indices.begin()), indices.end(),
+                           step_mismatch) != indices.end()) {
+        return {};
+    }
+    return std::make_pair(base, step);
+}
+
+// Count of the elements in a linear space from \p a to \b with \p step.
+int compute_linspace_length(int a, int b, int step) {
+    int distance = b - a;
+    internal_assert(distance >= 0 && distance % step == 0)
+        << "Unexpected linear space.";
+    int num_hops = distance / step;
+    return 1 + num_hops;
+}
+
+struct DemandedLanes {
+    DemandedLanes(const std::vector<int> &indices_)
+        : indices(indices_) {
+    }
+
+    auto begin() const {
+        return indices.begin();
+    }
+    auto end() const {
+        return indices.end();
+    }
+    auto size() const {
+        return indices.size();
+    }
+    auto empty() const {
+        return indices.empty();
+    }
+    int operator[](int i) const {
+        return indices[i];
+    }
+
+    std::vector<int> indices;  // Do *NOT* mutate in-place!
+};
+
 class Deinterleaver : public IRGraphMutator {
 public:
-    Deinterleaver(int starting_lane, int lane_stride, int new_lanes, const Scope<> &lets)
-        : starting_lane(starting_lane),
-          lane_stride(lane_stride),
-          new_lanes(new_lanes),
-          external_lets(lets) {
+    Deinterleaver(DemandedLanes lanes_, const Scope<> &lets)
+        : lanes(std::move(lanes_)), external_lets(lets) {
+    }
+
+    using IRGraphMutator::mutate;
+
+    Expr mutate(const Expr &e) override {
+        // FIXME: is there really only always-on assertions?
```

Comment:
Yes, I'd much rather have compiler bugs cause internal assertion failures than silent problems for users. We get a more useful bug report that way.

What are you checking here? Is the concern that the check is expensive?

Reply:
My concern is that this check is expensive and statically dead, it's not going to ever fire.